src/IR/types.ts

```diff
@@ -173,7 +191,12 @@ export function toString(a: Type): string {
 
 export function union(a: Type, b: Type): Type {
   try {
-    if (a.kind === "List" && b.kind === "List") {
+    if (a.kind === "Function" && b.kind === "Function") {
+      return functionType(
+        a.arguments.map((t, i) => union(t, b.arguments[i])),
```

Comment:
Function types should be contravariant in the parameter types, so this should be `intersect`, not `union`. But it's more work to implement `intersect`, so you might want to just throw an error if the parameter types differ.